dj/sql/parsing/ast.py

```diff
@@ -0,0 +1,491 @@
+        """
+        return self.filter(lambda n: isinstance(n, node_type))
+
+    def apply(self, func: Callable[["Node"], None]):
+        """
+        traverse ast and apply func to each Node
+        """
+        func(self)
+        for child in self.children:
+            child.apply(func)
+
+    def compare(self, other: "Node") -> bool:
+        """
+        compare two ASTs
+        """
+
+        return self == other and all(
+            child.compare(other_child)
+            for child, other_child in zip_longest(self.children, other.children)
+        )
+
+    def __eq__(self, other) -> bool:
+        """
+        Compares two nodes for "top level" equality.
+        Checks for type equality and primitive field types for full equality.
+        Compares all others for type equality only. No recursing.
+        Note: Does not check (sub)AST. See `Node.compare` for comparing (sub)ASTs.
+        """
+        primitives = {int, float, str, bool, type(None)}
+        return type(self) == type(other) and all(  # pylint: disable=C0123
+            s == o
+            if type(s) in primitives  # pylint: disable=C0123
+            else type(s) == type(o)  # pylint: disable=C0123
+            for s, o in zip(
+                (self.fields(False, False, False, True)),
+                (other.fields(False, False, False, True)),
+            )
+        )
+
+    @abstractmethod
+    def __hash__(self) -> int:
+        """
+        hash a node
+        """
+
+
+class Expression(Node):
+    """an expression type simply for type checking"""
+
+
+@dataclass(eq=False)  # type: ignore
+class Named(Expression):
+    """An Expression that has a name"""
+
+    name: str
+    quote_style: Optional[str]
+
+    @property
+    def quoted_name(self) -> str:
+        """
+        get the name of the Named Node including the quotes if any
+        """
+        return f'{self.quote_style if self.quote_style else ""}{self.name}{self.quote_style if self.quote_style else ""}'  # pylint: disable=C0301
+
+    def alias_or_name(self) -> str:
+        """
+        get the name or alias of the node
+        """
+        if len(self.parents) == 1:
+            parent = tuple(self.parents)[0]
+            if isinstance(parent, Alias):
+                return parent.name
+        return self.name
+
+
+class Operation(Expression):
+    """a type to overarch types that operate on other expressions"""
+
+
+class UnaryOpKind(Enum):
+    """the accepted unary operations"""
+
+    Plus = "+"  # pylint: disable=C0103
+    Minus = "-"  # pylint: disable=C0103
+    Not = "NOT"  # pylint: disable=C0103
+
+
+@dataclass(eq=False)
+class UnaryOp(Operation):
+    """an operation that operates on a single expression"""
+
+    op: UnaryOpKind  # pylint: disable=C0103
+    expr: Expression
+
+    def __hash__(self) -> int:
+        return hash((UnaryOp, self.op))
+
+
+class BinaryOpKind(Enum):
+    """the accepted binary operations"""
+
```

Comment:
It doesn't matter much but I think you can do `# pylint: disable=C0103` on this line and then `# pylint: enable=C0103` on line 252 to apply it over this range of lines.

Reply:
yeah much cleaner. Thank you!